### ISSUE TITLE
TypeError: Cannot read property 'fulfilled' of undefined

### ISSUE THREAD

**Issue by thoughtassassin** [jquense/yup#676]
**Describe the bug**
I have a Codesandbox running a library called React Hook Form. If you add two friends in the code pen and then remove the first friend and then submit the form you see the error. The stack trace shows the error at a line in react-hook-form.js. BUT if you dig deeper the error is actually coming from line 84 of runValidations.js where the check for !r.fulfilled throws an error saying:

`"TypeError: Cannot read property 'fulfilled' of undefined
    at eval (eval at <anonymous> (https://yycrz.csb.app/node_modules/yup/lib/util/runValidations.js), <anonymous>:1:3)
    at eval (https://yycrz.csb.app/node_modules/yup/lib/util/runValidations.js:84:17)
    at Array.filter (<anonymous>)
    at eval (https://yycrz.csb.app/node_modules/yup/lib/util/runValidations.js:83:32)
    at async Promise.all (index 0)
    at async validateWithSchema (https://yycrz.csb.app/node_modules/react-hook-form/dist/react-hook-form.js:383:21)
    at async eval (https://yycrz.csb.app/node_modules/react-hook-form/dist/react-hook-form.js:886:28)"`

**To Reproduce**
https://codesandbox.io/s/react-hook-form-array-fields-validation-using-refs-yycrz?from-embed

Add two friends and then delete one friend. When you submit the form, you can see the error Chrome dev tools.

> NOTE: if you do not provide a runnable reproduction the chances of getting feedback are significantly lower

**Expected behavior**
Handle the scenario where r is undefined to fail gracefully so it does not break the running script.

**Platform (please complete the following information):**

**Additional context**
Also, one of the folks at React Hook Form asked me to share this screenshot:
<img width="730" alt="1b0f20e8-8e5b-4b01-ac56-5ab844645213-Screen%20Shot%202019-10-26%20at%208 52 04%20am (1)" src="https://user-images.githubusercontent.com/1539057/67612988-b879d980-f76d-11e9-9a2d-39463cda6200.png">

**Comment by sijad** [comment]
I faced same issue:

```
Unhandled Rejection (TypeError): r is undefined
```

which refer to this line: https://github.com/jquense/yup/blob/0c27f076668632912311226259745d19129bf9bb/src/util/runValidations.js#L62

**Comment by VanTanev** [comment]
In some rare circumstances, you could work around this by adding [array.compact](https://github.com/jquense/yup#arraycompactrejector-value--boolean-schema) to your schema. 

#### Keep in mind this will not actually omit the empty values from the output of formik/react-hook-form/etc. 

However, it might be useful if you're using yup directly, or in circumstances where you can combine it with something like `.min(x).max(x)`, where `x` is the same number for both `min()` and `max()`.

**Comment by Nightliss** [comment]
I'm having the same issue, removing first element breaks validation. I'm using React Hook Form.

Adding check here
https://github.com/jquense/yup/blob/0c27f076668632912311226259745d19129bf9bb/src/util/runValidations.js#L62

`.filter(r => r ? !r.fulfilled : false`

fixes it for me locally. Hope it will be fixed.

**Comment by WonderPanda** [comment]
Also running into this using React Hook Form :(

**Comment by WonderPanda** [comment]
@Nightliss I got this working. I actually realized I was using the `register` function improperly when attaching refs to my input elements which was causing things to explode when removing elements.

I was passing it in with the format `ref={register}` instead of `ref={register({})`

**Comment by bsunderhus** [comment]
Is there any update in this issue?

**Comment by c0d3x** [comment]
I have same issue (also using React Hooks Form).
In my case it happens when using additional sub inputs with checkboxes that reference to index of mapped out checkboxes.
When I submit form and validation is triggered, it does not like that my array starts from 3 or 5 etc.. when array 0 is undefined I get this:
"Cannot read property 'fulfilled' of undefined"

Here is a codesandbox showing my issue, if you submit first checkbox it works, if you submit any of the other not having index in numerical order it fails:
https://codesandbox.io/s/react-hook-form-checkbox-array-with-inputs-g6pd4?from-embed

**Comment by arpit016** [comment]
is there any solution to this issue yet?

**Comment by sdwvit** [comment]
Hallo, I'm bumping into this issue. Using a mix of array.of, when, min-max and mixed. Not sure what causes problem, also have no desire to investigate. Please fix

**Comment by jquense** [comment]
Hi ya'll, if someone wants to post a runnable repro using the code sandbox template that demonstrates the problem **without** using a form library i can troubleshoot what is wrong. I don't have the bandwidth to provide support for each of these form libraries as well as yup

**Comment by sdwvit** [comment]
@jquense I'm going to find a workaround today, will post here. May give a clue to how to fix it

**Comment by sdwvit** [comment]
@jquense I fixed the issue. The reason is that formik was tracking custom fields which were not on the schema and which were used in Yup

**Comment by jquense** [comment]
happy to add better error handling but i don't know where this error is being thrown or why.

**Comment by VanTanev** [comment]
@jquense Here is a super simple reproduction code https://codesandbox.io/s/modest-spence-fn61p

**Comment by VanTanev** [comment]
@jquense So, this is the source of the bug:
https://github.com/jquense/yup/blob/27b287ba0dd8981ddb330c28123b41d85c4c8853/src/util/runValidations.js#L38-L49

Here, `promises.map` is supposed to map fulfilled and rejected promises to the appropriate value object. However, `Array.map` does not iterate over the empty spots in sparse arrays:

![image](https://user-images.githubusercontent.com/202527/85545149-5935d700-b624-11ea-9aed-bff0e65c9f39.png)

It can probably be said that the error originates here:
https://github.com/jquense/yup/blob/27b287ba0dd8981ddb330c28123b41d85c4c8853/src/array.js#L87

Here, no validation will be generated for empty Array positions.

---

It seems to me that for the purposes of this library, it makes sense to treat empty array positions as `undefined`..

**Comment by sdwvit** [comment]
well done @VanTanev

**Comment by sdwvit** [comment]
here a stacktrace I managed to get with this error, maybe it helps
![screencapture-localhost-3000-asset-configuration-manager-ac8b407d-05b1-ea11-86cd-00155d55771b-edit-2020-06-26-00_36_11](https://user-images.githubusercontent.com/8777372/85820835-25060780-b745-11ea-87fa-5bc305def4a3.png)

**Comment by VanTanev** [comment]
@sdwvit there is a fix in #950.

In the mean time, what is most probably happening is that you have an array notation in a Field name, eg:

```jsx
<Field name="user.friends[3].name" />
```

You can work around the problem by setting the initial values of that array manually to undefined:

```jsx
<Formik initialValues={{ 
  user: { friends: [undefined, undefined, undefined] } 
}} 
```

More generally, every time you deal with a form field name that accesses an array position, make sure that array position actually exists in the form's `initialValues` by doing `new Array(5).fill(undefined)`.

**Comment by arkniazi** [comment]
I'm still facing the same issue. Using @VanTanev suggestion solved the error.
